File: Day138_2025/Q138_Leetcode.py
class Solution:
    def colorTheGrid(self, m: int, n: int) -> int:
        from collections import defaultdict

        mod = 10**9 + 7
        valid = {}
        mask_end = 3 ** m

        # Generate valid column patterns
        for mask in range(mask_end):
            color = []
            mm = mask
            for _ in range(m):
                color.append(mm % 3)
                mm //= 3
            if all(color[i] != color[i + 1] for i in range(m - 1)):
                valid[mask] = color

        # Build compatible mask map
        adjacent = defaultdict(list)
        for mask1, color1 in valid.items():
            for mask2, color2 in valid.items():
                if all(color1[i] != color2[i] for i in range(m)):
                    adjacent[mask1].append(mask2)

        f = [0] * mask_end
        for mask in valid:
            f[mask] = 1

        for _ in range(1, n):
            g = [0] * mask_end
            for mask2 in valid:
                for mask1 in adjacent[mask2]:
                    g[mask2] = (g[mask2] + f[mask1]) % mod
            f = g

        return sum(f) % mod

File: Day138_2025/Q138_Leetcode.java
import java.util.*;

class Q138_Leetcode {
    int mod = 1_000_000_007;

    public int colorTheGrid(int m, int n) {
        Map<Integer, int[]> valid = new HashMap<>();
        int mask_end = (int)Math.pow(3, m);

        // Generate all valid column states
        for (int mask = 0; mask < mask_end; mask++) {
            int[] color = new int[m];
            int temp = mask;
            boolean isValid = true;
            for (int i = 0; i < m; i++) {
                color[i] = temp % 3;
                temp /= 3;
                if (i > 0 && color[i] == color[i - 1]) {
                    isValid = false;
                    break;
                }
            }
            if (isValid) valid.put(mask, color);
        }

        // Build adjacency map
        Map<Integer, List<Integer>> adjacent = new HashMap<>();
        for (var e1 : valid.entrySet()) {
            int mask1 = e1.getKey();
            int[] c1 = e1.getValue();
            for (var e2 : valid.entrySet()) {
                int mask2 = e2.getKey();
                int[] c2 = e2.getValue();
                boolean compatible = true;
                for (int i = 0; i < m; i++) {
                    if (c1[i] == c2[i]) {
                        compatible = false;
                        break;
                    }
                }
                if (compatible) {
                    adjacent.computeIfAbsent(mask1, _ -> new ArrayList<>()).add(mask2);
                }
            }
        }

        // DP initialization
        int[] dp = new int[mask_end];
        for (int mask : valid.keySet()) dp[mask] = 1;

        // DP iteration for each column
        for (int col = 1; col < n; col++) {
            int[] newDp = new int[mask_end];
            for (int mask2 : valid.keySet()) {
                for (int mask1 : adjacent.getOrDefault(mask2, new ArrayList<>())) {
                    newDp[mask2] = (newDp[mask2] + dp[mask1]) % mod;
                }
            }
            dp = newDp;
        }

        // Sum up all ways
        int ans = 0;
        for (int val : dp) {
            ans = (ans + val) % mod;
        }
        return ans;
    }
}
